src/Grpc.AspNetCore.Server/Internal/HttpContextServerCallContext.cs

```diff
@@ -0,0 +1,135 @@
+﻿#region Copyright notice and license
+
+// Copyright 2019 The gRPC Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#endregion
+
+using System;
+using System.Net.Sockets;
+using System.Threading;
+using System.Threading.Tasks;
+using Grpc.Core;
+using Microsoft.AspNetCore.Http;
+
+namespace Grpc.AspNetCore.Server.Internal
+{
+    internal class HttpContextServerCallContext : ServerCallContext
+    {
+        private string _peer;
+        private Metadata _requestHeaders;
+        private Metadata _responseTrailers;
+
+        internal HttpContextServerCallContext(HttpContext httpContext)
+        {
+            HttpContext = httpContext;
+        }
+
+        internal HttpContext HttpContext { get; }
+
+        internal bool HasResponseTrailers => _responseTrailers == null;
+
+        protected override string MethodCore => HttpContext.Request.Path.Value;
+
+        protected override string HostCore => HttpContext.Request.Host.Value;
+
+        protected override string PeerCore
+        {
+            get
+            {
+                if (_peer == null)
+                {
+                    var connection = HttpContext.Connection;
+                    if (connection.RemoteIpAddress != null)
+                    {
+                        _peer = (connection.RemoteIpAddress.AddressFamily == AddressFamily.InterNetwork ? "ipv4:" : "ipv6:") + connection.RemoteIpAddress + ":" + connection.RemotePort;
+                    }
+                }
+
+                return _peer;
+            }
+        }
+
+        // TODO(JunTaoLuo, JamesNK): implement this
+        protected override DateTime DeadlineCore => throw new NotImplementedException();
+
+        protected override Metadata RequestHeadersCore
+        {
+            get
+            {
+                if (_requestHeaders == null)
+                {
+                    _requestHeaders = new Metadata();
```

Comment:
In general I'd be concerned about duplicating collections. I need to verify that the header dictionary is immutable at this point.

Reply:
Similar issue with ResponseTrailers.

Reply:
From what I can tell the HttpContext related types aren't designed to be thread safe, e.g. https://github.com/aspnet/AspNetCore/blob/7bd5297cfa672c797e2fcbfffa6dc048699abd32/src/Http/Http/src/Features/ResponseCookiesFeature.cs#L55-L67

Reply:
It's not really thread safety, I was more concerned that operations on one collection will not be propagated to the other. I think I can update the implementation to make sure they are in sync though.

Reply:
Modifying the request headers on the server seems kind of odd. The amount of effort to keep them in sync doesn't seem like it would be worth that scenario.

Reply:
I think this is safe albeit not quite efficient.